### PR TITLE
Add kafka success/failure callbacks

### PR DESCRIPTION
**Pull request by joequery**
While initial connection timeouts to Kafka using the `KafkaProducer` are blocking, individual Request timeouts to Kafka using the `KafkaProducer` are non-blocking. This means attempting to catch an exception where we send a message to Kafka will not work.

## Problem demonstration

Checkout the current develop branch (currently at `8f01950c6c48`)

    $ git checkout develop

Start everything up

    $ docker-compose down && docker-compose up --build -d

Create a rule if necessary. Observe the tweet stream processing by tailing the traptor logs

    $ tail -f logs/traptor.log

Look for "tweet sent to kafka" messages.

```
{"message": "Tweet sent to kafka: 805880500107677703", "logger": "traptor-track-0", "timestamp": "2016-12-05T21:04:18.782111Z", "level": "INFO"}
```

Observe the current source that initiates this logging statement:

```python
try:
    self.kafka_conn.send(self.kafka_topic, enriched_data)
    self.logger.info("Tweet sent to kafka: {}".format(tweet.get('id_str', None)))
    dd_monitoring.increment('tweet_to_kafka_success')
except Exception:
    self.logger.error("Unable to add tweet to Kafka: {}".format(traceback.format_exc()))
    dd_monitoring.increment('tweet_to_kafka_failure')
```

The intent is that if the tweet was successfully sent to Kafka, we will see the "Tweet sent to kafka" message, and if we did not send the tweet to Kafka successfully, we will see "Unable to add tweet to Kafka". But the desired behavior is not matching the exhibited behavior.

To verify this, stop the kafka container (or process if running natively)

    $ docker-compose stop kafka

If the Exception was actually catching failed Kafka requests, we would see the "Unable to add tweet to Kafka" log message. Bare minimum, if the requests were timing out, we would not see incoming "Tweet sent to Kafka" messages. But after stopping Kafka the success message will still continue to display

This behavior has emerged since we have transitioned to the asynchronous `KafkaProducer` class - an upgrade from the `SimpleProducer` class. 

## Solution

Success and error actions should be defined using the `future` object returned by `self.kafka_conn.send`. The object allows us to define callback and callback-error functions.

### Solution verification

Stop the containers

    $ docker-compose down

Check out this branch

    $ git checkout --track origin/kafka_callbacks

Repeat the process we did for the `develop` branch - get tweets initially sending to Kafka successfully, then stop Kafka. You will (correctly) no longer see any "Tweet sent to Kafka" messages, as Kafka is not reachable. After some time, you will begin to see 

```
{"message": "Unable to add tweet to Kafka: None\n", "logger": "traptor-track-0", "timestamp": "2016-12-05T21:24:15.331229Z", "level": "ERROR"}
{"message": "KAFKA FAILURE: KafkaTimeoutError: ('Batch containing %s record(s) expired due to timeout while requesting metadata from brokers for %s', 3, TopicPartition(topic='traptor.testing', partition=0))", "logger": "traptor-track-0", "timestamp": "2016-12-05T21:24:15.332042Z", "level": "ERROR"}
```

Which means we are now correctly reporting failures

### CODE REVIEW

**Comment by madisonb** [comment]
I would be interested in knowing the runtime speed of this vs the `flush()`. I presume the flush is going to be slower but given the very high velocity we need traptor to run I don't want to implement a slower `futures` update unless we test it and it is marginally the same. Even running a simple set of timed for loops I think would suffice here instead of doing the testing directly within traptor.

File: traptor/traptor.py [hunk 1]
```diff
@@ -155,6 +157,19 @@ def _setup_kafka(self):
             self.logger.debug('Kafka_enabled setting: {}'.format(self.kafka_enabled))
             self.kafka_conn = None
 
+    def _gen_kafka_success(self):
+        def kafka_success(tweet, response):
+            self.logger.info("Tweet sent to kafka: {}".format(tweet.get('id_str', None)))
+            dd_monitoring.increment('tweet_to_kafka_success')
```
Comment (madisonb) [inline traptor/traptor.py]: We should put the `id_str` into the `extras`, can't do aggregations on the message string

File: traptor/traptor.py [hunk 2]
```diff
@@ -155,6 +157,19 @@ def _setup_kafka(self):
...
+        return kafka_success
+
+    def _gen_kafka_failure(self):
+        def kafka_failure(e):
+            self.logger.error("KAFKA FAILURE: {}".format(e))
+            self.logger.error("Unable to add tweet to Kafka: {}".format(traceback.format_exc()))
```
Comment (madisonb) [inline traptor/traptor.py]: Both tracebacks here can be condensed probably into a single exception message with the tracebacks inside the `extras` dict as `ex` or `ex_kafka`

File: traptor/traptor.py [hunk 3]
```diff
@@ -593,11 +608,12 @@ def _main_loop(self):
 
                     if self.kafka_enabled == 'true':
                         try:
-                            self.kafka_conn.send(self.kafka_topic, enriched_data)
-                            self.logger.info("Tweet sent to kafka: {}".format(tweet.get('id_str', None)))
-                            dd_monitoring.increment('tweet_to_kafka_success')
+                            self.logger.info("Attempting to send tweet {} to kafka".format(tweet.get('id_str', None)))
+                            future = self.kafka_conn.send(self.kafka_topic, enriched_data)
```
Comment (madisonb) [inline traptor/traptor.py]: Same thing here, we should remove the `id_str` as we can't do aggs

File: traptor/traptor.py [hunk 4]
```diff
@@ -593,11 +608,12 @@ def _main_loop(self):
...
+                            future.add_callback(self.kafka_success_callback, tweet)
+                            future.add_errback(self.kafka_failure_callback)
                         except Exception:
-                            self.logger.error("Unable to add tweet to Kafka: {}".format(traceback.format_exc()))
+                            self.logger.error("Uncaught exception sending tweet to Kafka: {}".format(traceback.format_exc()))
                             dd_monitoring.increment('tweet_to_kafka_failure')
```
Comment (madisonb) [inline traptor/traptor.py]: Is this exception catch even needed anymore since the futures handle things?

**Comment by madisonb** [comment]
@jasonrhaas or @rdempsey can one of you take point on tidying up the code in here? My curiosity got the better of me and I wrote the timer function myself. the futures work much nicer than the flushes as I expected.

```
t1() avg 0.0409413504601 flush()
t2() avg 0.000296490192413 futures
```
where t1 and t2 are from the following script
```
# not perfect but gets the job done
from kafka import KafkaProducer
import time

k = KafkaProducer(bootstrap_servers='scdev', # change me to your kafka server
                    reconnect_backoff_ms=4000,
                    retries=3,
                    linger_ms=25,
                    buffer_memory=4 * 1024 * 1024)

future_tally = 0

def success(self, future):
    global future_tally
    future_tally += 1
    pass

def failure(self, future):
    print "failure", future
    global future_tally
    future_tally += 1

def t1():
    result = k.send('sometopic', 'hey yo')
    k.flush()
    assert result.is_done

def t2():
    future = k.send('sometopic', 'blarg')
    future.add_errback(failure, future)
    future.add_callback(success, future)

sum = 0
t = time.time()
for _ in range(100):
    t1()
print "t1() avg", (time.time() - t) / 100, "flush()"

sum = 0
t = time.time()
for _ in range(100):
    t2()
while future_tally < 100:
    pass

print "t2() avg", (time.time() - t) / 100, "futures"
``` 
We just have to be careful here that we actually call the producer `close()` function (in this script the producer would not be killed cleanly, but I didnt want to take that into account when calculating just the time to producer correctly).

I think this needs to propagate throughout any code that uses the newer async kafka setup.

File: traptor/traptor.py (hunk 4 above)
Comment (joequery) [inline traptor/traptor.py]: Yes. send throws on initial connection failures and when the memory buffer is full. See the send method in the Kafka Python client in kafka.producers.kafka

Comment (joequery) [inline traptor/traptor.py]: https://github.com/dpkp/kafka-python/blob/9b59c5d755af73c2e7863e98b84b5882c297afda/kafka/producer/kafka.py#L472

File: traptor/traptor.py (hunk 1 above)
Comment (rdempsey) [inline traptor/traptor.py]: Updated.

File: traptor/traptor.py (hunk 2 above)
Comment (rdempsey) [inline traptor/traptor.py]: Done.

File: traptor/traptor.py (hunk 3 above)
Comment (rdempsey) [inline traptor/traptor.py]: Done.

File: traptor/traptor.py (hunk 4 above)
Comment (rdempsey) [inline traptor/traptor.py]: @joequery with the `def _gen_kafka_failure` message performing an increment of the `tweet_to_kafka_failure` in DD, is it needed in the Exception clause?

File: traptor/traptor.py [hunk 5]
```diff
@@ -593,11 +608,12 @@ def _main_loop(self):
 
                     if self.kafka_enabled == 'true':
                         try:
-                            self.kafka_conn.send(self.kafka_topic, enriched_data)
-                            self.logger.info("Tweet sent to kafka: {}".format(tweet.get('id_str', None)))
-                            dd_monitoring.increment('tweet_to_kafka_success')
+                            self.logger.info("Attempting to send tweet {} to kafka".format(tweet.get('id_str', None)))
+                            future = self.kafka_conn.send(self.kafka_topic, enriched_data)
+                            future.add_callback(self.kafka_success_callback, tweet)
```
Comment (rdempsey) [inline traptor/traptor.py]: @joequery I think `tweet` here needs to be replaced with the enriched_data object which is the tweet with the additional Traptor fields.

Comment (joequery) [inline traptor/traptor.py]: The tweet in the success callback is only used to yet the id. we can pass in the id directly instead if you'd like.

**Comment by jasonrhaas** [comment]
So I'm trying to duplicate this as well and I can't even get it to write to kafka with the current `develop` branch.

```
{"message": "Caught exception adding Twitter message to Kafka", "ex": "Traceback (most recent call last):\n  File \"traptor/traptor.py\", line 655, in _main_loop\n    self.kafka_conn.send(self.kafka_topic, enriched_data)\n  File \"/usr/local/lib/python2.7/site-packages/kafka/producer/kafka.py\", line 481, in send\n    self._wait_on_metadata(topic, self.config['max_block_ms'] / 1000.0)\n  File \"/usr/local/lib/python2.7/site-packages/kafka/producer/kafka.py\", line 598, in _wait_on_metadata\n    \"Failed to update metadata after %s secs.\", max_wait)\nKafkaTimeoutError: KafkaTimeoutError: ('Failed to update metadata after %s secs.', 60.0)\n", "logger": "traptor-track-0", "timestamp": "2016-12-14T08:52:17.971171Z", "level": "ERROR"}
```

**Comment by jasonrhaas** [comment]
OK I figured it out.  The KAFKA_ADVERTISED_HOSTNAME was breaking things.  I removed that from the docker-compose config and it fixed the problem.  wtf

```
  kafka:
    image: wurstmeister/kafka
    ports:
      - "9092:9092"
    environment:
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
      # KAFKA_CREATE_TOPICS: "test:1:1"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    depends_on:
      - zookeeper
```

**Comment by jasonrhaas** [comment]
This is sweet!  Finally got it working... does this mean we don't need the little bounded bugger thing that @andrewkcarter wrote?  It appears like the kafka producer buffer does indeed work pretty well.

image

**Comment by jasonrhaas** [comment]
image

In the image above you can see what happens.  The green is the "Tweet send to Kafka".  Light purple is the "Attempts", and Dark Purpose is the "Kafka errors".  When I shut down kafka you can see the errors pile up.  When I turn it back on you can see the big spike in tweets sent and then it goes back to steady state mode.

**Comment by joequery** [comment]
I do believe this removes the need for the async queue that Andrew suggested previously.

**Comment by rdempsey** [comment]
@jasonrhaas since you tested the PR do you want to approve it and merge it into dev?

**Comment by jasonrhaas** [comment]
Merging...

**Pull request merged**